### PR TITLE
drivers: adc: stm32h7: Oversampling Ratio set incorrectly

### LINKED ISSUE #38606: drivers: adc: stm32h7: Oversampling only supported for ADC3
**Describe the bug**
This [PR](https://github.com/zephyrproject-rtos/zephyr/pull/36714/) provided the ability to do ADC oversampling, which then also added support for the H7 in this [PR](https://github.com/zephyrproject-rtos/zephyr/pull/37392). This will however not set the ratio correctly.

According to the [Zephyr ADC docs](https://docs.zephyrproject.org/latest/reference/peripherals/adc.html#c.adc_sequence.oversampling) the value `sequence->oversampling` should be `2^N`. This means if you require a ratio of 256 you would define `sequence->oversampling=8`. But using ADC1 or ADC2 with the current code if you set `sequence->oversampling=8` it will result in a ratio of 7, and not 256. It requires the ratio in a different format than for ADC3 for which it seems this driver was written.

**For ADC1 and ADC2:**
> OSVR[9:0]: Oversampling ratio
This bitfield is set and cleared by software to define the oversampling ratio.
0: 1x (no oversampling)
1: 2x
2: 3x
...
1023: 1024x

**For ADC3:** 
> OVSR[2:0]: Oversampling ratio
This bitfield is set and cleared by software to define the oversampling ratio.
000: 2x
001: 4x
010: 8x
011: 16x
100: 32x
101: 64x
110: 128x
111: 256x

If using ADC1 or ADC2 I would recommend simply stating:
```
uint16_t oversampling_ratio = (1<<sequence->oversampling) - 1;
...
LL_ADC_ConfigOverSamplingRatioShift(adc, oversampling_ratio, ...);
```

### PR DESCRIPTION
**Pull request by FRASTM**
Configure the with Oversampling ratio for the stm32H7 soc devices, with the LL_ADC_ConfigOverSamplingRatioShift
The LL API function dffers from other stm32 soc : 
ratio is directly the value to set from 1..1024
Depending on the ADC instance ratio is directly the value to set from 1..1024   --> ADC of the stm32H74x for example
or is directly a 3-bit   --> **ADC3** of the stm32H723 for example

requires https://github.com/zephyrproject-rtos/hal_stm32/pull/117

fixes https://github.com/zephyrproject-rtos/zephyr/issues/38606

Signed-off-by: Francois Ramu francois.ramu@st.com

### CODE REVIEW

**Comment by heinwessels** [comment]
@FRASTM as mentioned in [#38606](https://github.com/zephyrproject-rtos/zephyr/issues/38606) this code will not handle `ADC3` correctly for H7. Even if the [HAL Bug](https://github.com/STMicroelectronics/STM32CubeH7/issues/177) is solved, ADC3 still requires a 3-bit value, and your code will send a `10bit` value. It will also not handle the `2^N` input number correctly that Zephyr specifies in it's documentation.

**Comment by FRASTM** [comment]
include the commit for the hal_stm32 : west.yml: update modules/hal/stm32/hal_stm32 for stm32h7

**Comment by erwango** [comment]
@heinwessels Would you mind having a new look ?

**Comment by heinwessels** [comment]
> @heinwessels Would you mind having a new look ?

I'm AFK for the next two weeks so I cannot test the code. I did find a possible issue in HAL PR #117 that this PR depends on, although this PR looks good.

**Comment by FRASTM** [comment]
Keeping the` LL_ADC_ConfigOverSamplingRatioShift(adc, sequence->oversampling`   for the stm32H7
It means that the 'Ratio' in the LL function is a 'n' value from 1 to 8 (ADC3) or from 1 to 10 (other ADC)  and no more the corresponding to a oversampling of 2^n. 
That makes the adc_stm32_start_conversion() unchanged for the oversampling configuration (excepting a possible extension to case 9 and case 10) but the change is reported to the stm32cube LL function.

File: drivers/adc/adc_stm32.c [hunk 1]
```diff
@@ -386,7 +386,7 @@ static int start_read(const struct device *dev,
 	 * This register controls the analog switch integrated in the IO level.
 	 * NOTE: There is no LL API to control this register yet.
 	 */
-	adc->PCSEL |= channels & ADC_PCSEL_PCSEL_Msk;
+	LL_ADC_SetChannelPreSelection(adc, index);
```
Comment (ABOSTM) [inline drivers/adc/adc_stm32.c:389]: Isn't it `channel `instead of `index `?

**Review by ABOSTM — CHANGES_REQUESTED** [review]

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by heinwessels — APPROVED** [review]
LGTM, atthough I cannot test the code at the moment. 

Also, I tend to agree with ABOSTM's [comment](https://github.com/zephyrproject-rtos/hal_stm32/pull/117#discussion_r720037396) on the HAL Driver change, which would mean this driver will need to be updated.

**Comment by FRASTM** [comment]
rebase

File: drivers/adc/adc_stm32.c [hunk 2]
```diff
@@ -386,7 +386,7 @@ static int start_read(const struct device *dev,
 	 * This register controls the analog switch integrated in the IO level.
 	 * NOTE: There is no LL API to control this register yet.
```
Comment (heinwessels) [inline drivers/adc/adc_stm32.c:387]: This comment is outdated.

**Comment by FRASTM** [comment]
New version that take the new LL_ADC_ConfigOverSamplingRatioShift stm32Cube function
where the Ratio is a value from 1..1024 (or a 2^n value from 1..256 for ADC3)
This means that the parameter to give to the LL_ADC_ConfigOverSamplingRatioShift is 2^oversampling (and no more the 'oversampling').

**Review by ABOSTM — APPROVED** [review]
LGTM

**Review by erwango — CHANGES_REQUESTED** [review]

**Review by erwango — CHANGES_REQUESTED** [review]
One minor comment, otherwise LGTM

**Review by erwango — APPROVED** [review]

**Pull request merged**